### PR TITLE
setCookie throws 500 error instead of 401 when token is invalid/expired

### PR DESCRIPTION
**Pull request by akarukappadath**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [x] I've documented my API changes in Swagger

In all cases:

- [x] Get two thumbsworth of review and PO signoff if necessary
- [x] Verify all tests go green
- [x] Squash and merge; you can delete your branch after this
- [x] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/Leonardo.scala [hunk 1]
```diff
@@ -177,6 +177,12 @@ object Leonardo extends RestClient with LazyLogging {
       val cookie = Cookie(HttpCookiePair("LeoToken", token.value))
       handleContentItemResponse(parseResponse(getRequest(url + path, httpHeaders = List(cookie))))
     }
+
+    def setCookie(googleProject: GoogleProject, clusterName: ClusterName)(implicit token: AuthToken, webDriver: WebDriver): String = {
+      val path = notebooksPath(googleProject, clusterName) + "/setCookie"
+      logger.info(s"Set cookie: GET /$path")
+      parseResponse(getRequest(url + path))
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/Leonardo.scala]: So this sends a `/setCookie` request _without_ a token, which should always return 403. I'm wondering if this method should actually pass the token that it receives. The test below could explicitly pass it a bad token. It makes this method more generic and reusable.

I think this should work:
```
parseResponse(getRequest(url + path, httpHeaders = List(Authorization(OAuth2BearerToken(token.value)))))
```

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 2]
```diff
@@ -419,5 +419,13 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
         }
       }
     }
+
+    "should throw 401 for invalid token" in withWebDriver { implicit driver =>
+      try {
+        Leonardo.notebooks.setCookie(ronCluster.googleProject, ronCluster.clusterName)
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: Per my above comment you could explicitly pass this method a bad token like:
```
Leonardo.notebooks.setCookie(ronCluster.googleProject, ronCluster.clusterName)(someBadAuthToken, driver)
```

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 3]
```diff
@@ -419,5 +419,13 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
...
+      } catch {
+        case e: Exception => e.getMessage should include(""""statusCode":409""")
+      }
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: So here you're correctly expecting the `setCookie` call to fail, and verifying the error message. But if the call to `/setCookie` unexpectedly _succeeds_, then the test will pass. A pattern for negative testing is often the following:

```
try {
  // call thing expected to fail
  fail("Expected thing to fail, but got success")
} catch {
  case e: Exception => // verify exception
}
```
ScalaTest actually provides a different API for doing this. Here is an example:
https://github.com/DataBiosphere/leonardo/blob/develop/automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala#L113-L119

I would recommend doing something like the above link.

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleDataprocDAO.scala [hunk 4]
```diff
@@ -187,7 +188,14 @@ class HttpGoogleDataprocDAO(appName: String,
     val request = oauth2.tokeninfo().setAccessToken(accessToken)
     retryWhen500orGoogleError(() => executeGoogleRequest(request)).map { tokenInfo =>
       (UserInfo(OAuth2BearerToken(accessToken), WorkbenchUserId(tokenInfo.getUserId), WorkbenchEmail(tokenInfo.getEmail), tokenInfo.getExpiresIn.toInt), Instant.now().plusSeconds(tokenInfo.getExpiresIn.toInt))
-    }.handleGoogleException(GoogleProject(""), Some("oauth"))
+    } recover {
+    case e: GoogleJsonResponseException =>
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleDataprocDAO.scala]: Missing indentation here

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleDataprocDAO.scala [hunk 5]
```diff
@@ -187,7 +188,14 @@ class HttpGoogleDataprocDAO(appName: String,
...
+      val msg = s"Call to Google OAuth API failed. Status: ${e.getStatusCode}. Message: ${e.getDetails.getMessage}"
+      logger.error(msg, e)
+      throw new WorkbenchException(msg, e)
+    case e: IllegalArgumentException =>
+      throw AuthorizationError()
```
Comment (rtitle) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/HttpGoogleDataprocDAO.scala]: I'd like to have a comment here so we understand the context. Something like:

// Google throws IllegalArgumentException when passed an invalid token. Handle this case and rethrow a 401.

Also include a link to the github ticket.

**Comment by akarukappadath** [comment]
jenkins retest

**Comment by rtitle** [comment]
The code looks good. There some git things to do though :)

1. `.DS_Store` and `automation/firecloud-automated-testing` are still showing up in this PR. I think these steps should remove them:
   ```
   // remove .DS_Store
   git rm .DS_Store
   add `.DS_Store` to `.gitignore` -- it should probably be in there anyway.
   ```
   ```
   // remove firecloud-automated-testing
   git rm -rf automation/firecloud-automated-testing
   remove submodule section from your .git/config
   rm -rf .git/modules
   ```
   I _think_ the above should remove the submodule. You probably mistakenly cloned https://github.com/broadinstitute/firecloud-automated-testing under Leo's automation directory. But this is a separate repo so it should be cloned as a sibling of leonardo.

   Then `git add --all`, `git commit -m ...`, and `git push`



2. It looks like the PR is complaining about conflicts. Once you've done the above, try to resolve them via:
   ```
   git pull
   git rebase -i origin/develop
   save the file to keep all your commits
   ```
   It may pop up a merge editor to resolve conflicts. We can walk through this together.

**Comment by akarukappadath** [comment]
jenkins retest

**Comment by akarukappadath** [comment]
jenkins retest

**Comment by rtitle** [comment]
Seems you got a failure in fiab-start. I think this was a known issue yesterday which has now been fixed. Hopefully another jenkins retest will make this pass.

**Comment by akarukappadath** [comment]
jenkins retest

**Comment by akarukappadath** [comment]
jenkins retest

**Pull request merged**